src/h3lib/lib/h3Index.c

```diff
@@ -779,6 +779,86 @@ void H3_EXPORT(h3ToGeoBoundary)(H3Index h3, GeoBoundary* gb) {
                           H3_EXPORT(h3IsPentagon)(h3), gb);
 }
 
+/**
+ * Returns the max number of possible icosahedron faces an H3 index
+ * may intersect.
+ *
+ * @return int count of faces
+ */
+int H3_EXPORT(maxFaceCount)(H3Index h3) {
+    // a pentagon always intersects 5 faces, a hexagon never intersects more
+    // than 2 (but may only intersect 1)
+    return H3_EXPORT(h3IsPentagon)(h3) ? 5 : 2;
+}
+
+/**
+ * Find all icosahedron faces intersected by a given  H3 index, represented
+ * as integers from 0-19. The array is sparse; since 0 is a valid value,
+ * invalid array values are represented as -1. It is the responsibility of
+ * the caller to filter out invalid values.
+ *
+ * @param h3 The H3 index
+ * @param out Output array. Must be of size maxFaceCount(h3).
+ */
+void H3_EXPORT(h3GetFaces)(H3Index h3, int* out) {
+    int res = H3_GET_RESOLUTION(h3);
+    int isPentagon = H3_EXPORT(h3IsPentagon)(h3);
+
+    // We can't use the vertex-based approach here for class II pentagons,
+    // because all their vertices are on the icosahedron edges. Their
+    // direct child pentagons cross the same faces, so use those instead.
+    if (isPentagon && !isResClassIII(res)) {
+        H3Index childPentagon = makeDirectChild(h3, 0);
+        H3_EXPORT(h3GetFaces)(childPentagon, out);
+        return;
+    }
+
+    // convert to FaceIJK
+    FaceIJK fijk;
+    _h3ToFaceIjk(h3, &fijk);
+
+    // Get all vertices as FaceIJK addresses. For simplicity, always
+    // initialize the array with 6 verts, ignoring the last one for pentagons
```

Comment:
Nit: This array isn't initialized, you've literally told the C compiler to leave the space for 6 structs directly allocated in the library for this function to use for this operation. (Which also makes this method not-parallel-computing-friendly.)

Reply:
Sorry, array handling is still the most confusing part for me of working in C. You're saying I should use `malloc` here?

Reply:
Or would this work with `FaceIJK fijkVerts[NUM_HEX_VERTS] = {0}`?

Reply:
The first one. The memory needs to be allocated either on the heap or on the stack. This syntax tells the compiler to put some dedicated space somewhere in the `.so/.dll` file and the function then references that space for all calls.

If multiple threads are running the same function at the "same time" (either actually at the same time on different cores, or one thread was running and a CPU interrupt happened partway through and the kernel scheduler spun up another thread and it continued in the same function) you could have very-hard-to-track bugs where the function gives you the wrong answer a vanishingly small amount of the time.

Reply:
Got it. I was expecting the array allocation here to work the same as allocating a struct (which doesn't do what you're describing, right?).

Reply:
It looks like the original line was `FaceIJK fijkVerts[NUM_HEX_VERTS];` This looks to me like a stack allocation - if it were `static FaceIJK fijkVerts[NUM_HEX_VERTS];` then it would be in the `bss` section in the library.

I think `FaceIJK fijkVerts[NUM_HEX_VERTS] = {0};` should be OK.

Reply:
![yes but no](https://i.ytimg.com/vi/IGXYzNm29bQ/hqdefault.jpg)

When you have a function signature like: `StructName fnName();` your program is going to copy the value of the struct into a new memory location and not be affected after-the-fact, but a variable declaration within the function like `StructName structVar;` or `StructName structVar = { /* struct contents */ };` *is* going to also be a statically-allocated blob in the `.so/.dll` file that could suffer the same interrupt issue across multiple threads.

But if that struct is just internal to the function (not part of the output signature) sometimes the C compiler/optmizer will break it into its constituent fields as registers on the CPU, and CPU interrupts properly store and retrieve the registers for you and you luck out. (Also why basic `int`/`float`/`double`/etc declarations don't cause problems usually, the variables get completely converted into registers or stack allocations/deallocations.) C is unfortunately low-level enough where this matters but "high-level" enough that you can't clearly declare your intentions, here.

Allocating these things on the stack or the heap is the safest way to go for parallel-friendly functions. (This is probably why @isaachier pushed for so much stack allocation stuff, since it's faster than heap and handles this situation, but unfortunately some platforms don't allow stack allocation and some of our functions can allocate way too much memory.)

I'm thinking H3 should have some [valgrind](http://valgrind.org/) testing set up to check for memory leaks and parallelization issues.

Reply:
> It looks like the original line was `FaceIJK fijkVerts[NUM_HEX_VERTS];` This looks to me like a stack allocation - if it were `static FaceIJK fijkVerts[NUM_HEX_VERTS];` then it would be in the `bss` section in the library and the reuse across function calls would be an issue.
> 
> I think `FaceIJK fijkVerts[NUM_HEX_VERTS] = {0};` should be OK.

I thought stack allocation was disabled because it was causing problems for the ARM builds?

Reply:
> I'm thinking H3 should have some valgrind testing set up to check for memory leaks and parallelization issues

+1, this is a good addition to our CI if possible.

Reply:
> I'm thinking H3 should have some [valgrind](http://valgrind.org/) testing set up to check for memory leaks and parallelization issues.

> +1, this is a good addition to our CI if possible.

We have an option for this in our CMake setup but we don't run it in CI right now. I think that would be a really good addition to automatically run it. 

> I thought stack allocation was disabled because it was causing problems for the ARM builds?

We removed dynamic stack allocations using [alloca](https://linux.die.net/man/3/alloca). We did not make any compiler option changes relating to how local variables are allocated AFAIK. I get the impression we might be thinking different compiler options are being applied here.

> When you have a function signature like: `StructName fnName();` your program is going to copy the value of the struct into a new memory location and not be affected after-the-fact, but a variable declaration within the function like `StructName structVar;` or `StructName structVar = { /* struct contents */ };` _is_ going to also be a statically-allocated blob in the `.so/.dll` file that could suffer the same interrupt issue across multiple threads.

I don't follow this. Could you send me an example program?

Reply:
> > I'm thinking H3 should have some [valgrind](http://valgrind.org/) testing set up to check for memory leaks and parallelization issues.
> 
> > +1, this is a good addition to our CI if possible.
> 
> We have an option for this in our CMake setup but we don't run it in CI right now. I think that would be a really good addition to automatically run it.
> 
> > I thought stack allocation was disabled because it was causing problems for the ARM builds?
> 
> We removed dynamic stack allocations using [alloca](https://linux.die.net/man/3/alloca). We did not make any compiler option changes relating to how local variables are allocated AFAIK. I get the impression we might be thinking different compiler options are being applied here.
> 

I was thinking of this issue: https://github.com/uber/h3/issues/233 so it was iOS-specific, not ARM-specific, but I remembered it involving a mobile platform. I thought we were using some of the compiler options to enforce that. I thought we were using something like `-mstack-increment=NUM` or `-fmax-stack-var-size` with a low value which would push these to become part of the data section and shared.

I was also confused because gcc says this about alloca:
```
       -msoft-stack
           Generate code that does not use ".local" memory directly for stack storage. Instead, a per-warp stack pointer is maintained explicitly. This enables variable-length stack allocation (with variable-length arrays or
           "alloca"), and when global memory is used for underlying storage, makes it possible to access automatic variables from other threads, or with atomic instructions. This code generation variant is used for OpenMP
           offloading, but the option is exposed on its own for the purpose of testing the compiler; to generate code suitable for linking into programs using OpenMP offloading, use option -mgomp.
```
Which implies that disabling variable-length stack allocation causes the stack to be a static block of shared memory.

> > When you have a function signature like: `StructName fnName();` your program is going to copy the value of the struct into a new memory location and not be affected after-the-fact, but a variable declaration within the function like `StructName structVar;` or `StructName structVar = { /* struct contents */ };` _is_ going to also be a statically-allocated blob in the `.so/.dll` file that could suffer the same interrupt issue across multiple threads.
> 
> I don't follow this. Could you send me an example program?

It would only apply if the earlier assumption is true: that the compiler was given a constraint on the amount of stack memory it can allocate for a given function and the struct in question was too large. It would have to switch from a stack allocation to fixed bytes in bss or data similar to how static variables work (and it would work fine for single-threaded applications).

Reply:
Just checking - this conversation is largely academic, right? I switched to `malloc`, so unless we see a potential downside to using `malloc` here I think this should be fine now.

Reply:
Mostly academic, yes, but if the compiler can do the allocation in stack space and not cause iOS to barf, that would be faster (and more predictable in speed) than malloc, which will sometimes have to call out to the kernel to get a new page of memory and cause an interrupt.

I would say the malloc is definitely safer with so many compile targets to worry about.

Reply:
I don't believe that 6 FaceIJK structures would be an issue, so I'd suggest using stack allocation here. I don't think we need to be quite that careful about stack space.

Reply:
Is `FaceIJK fijkVerts[NUM_HEX_VERTS] = {0}` using stack allocation, then? Is that the suggested change?

Reply:
Updated to `FaceIJK fijkVerts[NUM_HEX_VERTS] = {0}` per @isaac's suggestion.

Reply:
The previous formulation had compiler warnings, so I'm back to my original `FaceIJK fijkVerts[NUM_HEX_VERTS];` implementation. The internet seems to think this is a stack allocation, and @dfellis's original concern wasn't valid (unless this was declared at the file level and/or `static`). Please let me know if there are additional concerns here.